src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Updates project system service with new project info.
+/// </summary>
+/// <remarks>
+/// Used to figure out if the project system data is being added, updated or deleted
+/// and acts accordingly to update project system service.
+/// </remarks>
+internal class ProjectConfigurationStateManager
+{
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+    private readonly IRazorProjectService _projectService;
+    private readonly ILogger _logger;
+
+    /// <summary>
+    /// A set of all projects known to this service.
+    /// </summary>
+    private readonly HashSet<ProjectKey> _projectSet;
```

Comment:
Rather than this class having its own cache of project keys, I think it would be better for it to just use the ProjectSnapshotManager. The RazorProjectService hides that, and its right that this class uses the RazorProjectService, but there should be no issue in it also using the manager directly, to reason about which projects are known and not known.

Reply:
https://github.com/dotnet/razor/pull/10182/commits/e9e5caafeba534a78878e3053f653ea66e1e1f34